pkg/controller/deployment/deployment_controller.go

```diff
@@ -244,7 +259,7 @@ func (dc *DeploymentController) getDeploymentsForReplicaSet(rs *apps.ReplicaSet)
 	if len(deployments) > 1 {
 		// ControllerRef will ensure we don't do anything crazy, but more than one
 		// item in this list nevertheless constitutes user error.
-		klog.V(4).InfoS("user error! more than one deployment is selecting replica set",
+		klog.FromContext(ctx).V(4).Info("user error! more than one deployment is selecting replica set",
```

Comment:
I haven't looked a the specific functions, but would like to point out that there's a tradeoff: declaring a logger variable early makes sense when it then gets used multiple times. But it might still make sense to push it after some initial return statements, in particular when those include the most common path through a function, to avoid the (smallish) cost for the lookup when the logger is not needed.

If a function only needs a logger in an uncommon error path, then it is better to only look up the logger when needed.

Reply:
I agree for the (smallish) optimization to lookup the logger only when needed. 

My argument was in favor of readability and maintainability, so the logger is accessible from anywhere in the function for the future changes in this function, and there is no risk a second lookup is added to get the same logger in the same function.

Reply:
Yes, it's a tradeoff between performance and maintainability.